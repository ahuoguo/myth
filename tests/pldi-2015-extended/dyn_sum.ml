type nat =
   | O
   | S of nat

type dyn =
  | Error
  | Base of nat
  | Dyn of (dyn -> dyn)


let succ (d:dyn) : dyn =
  match d with
  | Error -> Error
  | Base ( n ) -> Base (S(n))
  | Dyn ( f ) -> Error
;;

let id (d:dyn) : dyn = d ;;

let dyn_sum : dyn -> dyn -> dyn |>
{ Dyn (id) =>
  ( Error    => Error
  | Dyn (id) => Error
  | Base (0) => Error
  | Base (1) => Error
  | Base (2) => Error )
| Error =>
  ( Error    => Error
  | Dyn (id) => Error
  | Base (0) => Error
  | Base (1) => Error
  | Base (2) => Error )
| Base (0) =>
  ( Error    => Error
  | Dyn (id) => Error
  | Base (0) => Base (0)
  | Base (1) => Base (1)
  | Base (2) => Base (2) )
| Base (1) =>
  ( Error    => Error
  | Dyn (id) => Error
  | Base (0) => Base (1)
  | Base (1) => Base (2)
  | Base (2) => Base (3) )
| Base (2) =>
  ( Error    => Error
  | Dyn (id) => Error
  | Base (0) => Base (2)
  | Base (1) => Base (3)
  | Base (2) => Base (4) )
} = ?
