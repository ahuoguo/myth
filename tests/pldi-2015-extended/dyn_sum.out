(* Generated Thu Nov 13 01:56:16 2014 *)
let dyn_sum : dyn -> dyn -> dyn =
  fun (d1:dyn) ->
    fun (d2:dyn) ->
      match d1 with
        | Error -> Error
        | Base (n1) -> (match n1 with
                          | O -> (match d2 with
                                    | Error -> Error
                                    | Base (n2) -> d2
                                    | Dyn (f3) -> Error)
                          | S (n2) -> (match n2 with
                                         | O -> succ d2
                                         | S (n3) -> succ (succ d2)))
        | Dyn (f3) -> Error
;;
